src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+                    };
+                    if dnssec_ok && !ad {
+                        ad = true;
+                    }
+
+                    let key =
+                        Key::new(qname, qclass, qtype, ad, cd, dnssec_ok, rd);
+                    let opt_ce = self.cache_lookup(&key).await?;
+                    if let Some(value) = opt_ce {
+                        let opt_response =
+                            handle_cache_value::<_, C>(qname, value);
+                        if let Some(response) = opt_response {
+                            return response;
+                        }
+                    }
+
+                    let request =
+                        self.upstream.send_request(self.request_msg.clone());
+                    self.state = RequestState::GetResponse(key, request);
+                    continue;
+                }
+                RequestState::GetResponse(key, request) => {
+                    let response = request.get_response().await;
+
+                    let key = key.clone();
+                    let value = Arc::new(Value::new(
+                        &response,
+                        &self.config,
+                        &self.clock,
+                    )?);
+                    self.cache_insert(key, value).await;
+
+                    return response;
+                }
+                RequestState::GetResponseNoCache(request) => {
+                    return request.get_response().await;
+                }
+            }
+        }
+    }
+
+    /// Try to find an cache entry for the key.
+    async fn cache_lookup(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // There are 4 flags that may affect the response to a query.
+        // In some cases the response to one value of a flag could be
+        // used for the ohter value.
+        // This function takes all 4 flags. First we fix the CD flag.
+        // This flag has to be used as is. Next we have the request to
+        // a function that looks at RD, DO, and AD.
+        self.cache_lookup_rd_do_ad(key).await
+    }
+
+    /// Try to find an cache entry for the key taking into account the
+    /// RD, DO, and AD flags.
+    async fn cache_lookup_rd_do_ad(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // For RD=1 we can only use responses to queries with RD set.
+        // For RD=0, first try with RD=0 and then try with RD=1. If
+        // RD=1 has an answer, store it as an answer for RD=0.
+        let opt_value = self.cache_lookup_do_ad(key).await?;
+        if opt_value.is_some() || key.rd {
+            return Ok(opt_value);
+        }
+
+        // Look if there is something with RD=1. We can use the
+        // response unmodified.
+        let mut alt_key = key.clone();
+        alt_key.rd = true;
+        let opt_value = self.cache_lookup_do_ad(&alt_key).await?;
+        if let Some(value) = &opt_value {
```

Comment:
This `if` ... `new_from_value_and_response()` ... `self.cache_insert()` pattern is repeated multiple times below with only the `clear_rd()` part changing as I can see, so if I were writing this I would factor out these common parts, but if possible but not convenient right now that can be left till later.